### ISSUE TITLE
Methods to grab hardware information about player's computer

### ISSUE THREAD

**Issue by KingofBeast** [Facepunch/garrysmod-requests#745]
For the sake of statistics, let Lua know hardware information about the computer it's running on.

`system.GPUInfo()`
`system.CPUInfo()`
`system.MotherboardInfo()`
`system.RAMInfo()`
`system.DisplayCount()`

You get the idea. Really as much harmless info we can get about our PC.

**Comment by aStonedPenguin** [comment]
maybe even something like system.GetHardwareInfo(). I would love to collect some statistics.

**Comment by Kefta** [comment]
Someone requested this to Garry before iirc, but he refused since he believed that info should be private.

**Comment by thegrb93** [comment]
Steam at least asks you if you want to participate in their surveys. This is going too far.

**Comment by gspetrou** [comment]
I'm somewhat against this because any server owner would be able to gather this information from their clients. At least the big name companies ask for permission first, even if the info is not that important.

**Comment by KingofBeast** [comment]
"Oh no! They know how much RAM I have! D:"

**Comment by Kefta** [comment]
I think a clientside convar would be the only way to ask for a user's "permission." Servers could check the value, then popup a short explanation on why they should enable it for statistics. Otherwise, I couldn't really think of a more logical solution.

You could always just make a module and ask members to install it, but that would really only work for large communities.

**Comment by aStonedPenguin** [comment]
Basic hardware info is pretty harmless.

**Comment by xbeastguyx** [comment]
Agreed, this would be really useful to help target weak points, as well as general statistics.

**Comment by Noiwex** [comment]
@Kefta i sure love em convars and checkboxes

**Comment by Noiwex** [comment]
This info tho could be used as hwid or something, and I suspect this is the main target of inferior (ops, superior) servers.

**Comment by Jvs34** [comment]
Considering this wouldn't really be useful for scripting other than info collection (except for maybe display count?), it'd be best to have all those functions condensed into one, as you really only need to call this stuff once anyway.

**Comment by kpjVideo** [comment]
DisplayCount and GPUInfo would be great! I dont really think it's too malicious unless it gave out, like others have said, specific information per user. (Allowing servers to target specific individuals)

**Comment by Kefta** [comment]
@Jvs34 I could see GPU info being useful as well in detecting whether to do certain operations (halos, setting model quality, etc), but that would be very specific and on a case-by-case basis.

**Comment by wyozi** [comment]
@Kefta you cannot realistically do that as there are so many GPU models available.

Having the functions be freely usable by everyone would definitely be detrimental to normal user privacy. I know I'd use the information for fingerprinting. If this was implemented there should be a Do-Not-Track option in the options menu (that cannot be modified by clientside Lua in any way), that blocks access to hardware info functions and is enabled by default.

**Comment by FPtje** [comment]
What does UK law say about this?

**Comment by KingofBeast** [comment]
@wyozi I don't think it'd be as daunting of a task as you think, with some simple string parsing to get the 690,750,980,290 etc out of the GPU name and line it up with some general performance values for each one. That said I think tracking average fps would be better for determining graphics optimizations than that since we can't account for other resource heavy tasks the computer may be doing.

@Noiwex whatever issues you have with popular servers, you are better off acting like an adult on a legitimate feature request. There's not always an ulterior motive for everything; I just like making graphs. Ask @Kefta

I'd also like to point out that the benefit of using this to fingerprint people is.. Almost none. Imagine how many people use an i7 4820k and GTX 980. Results would be completely unreliable with the exception of people with extreme toasters or super computers. It really is harmless info.

**Comment by DrVrej** [comment]
Dude don't add this, I don't want people to know how much RAM I have. This would break privacy bro please don't add. They can give me a virus if they how much RAM I have! Please don't add! They could even end up installing extra ram without my permission!

**Comment by Kefta** [comment]
@wyozi It's realistic: the primary thing you need to check is for Intel integrated graphics, which you could tell just by the brand string.

And the "Do-Not-Collect" option was something I suggested to KoB yesterday since that's the model Google follows for data collection.

@FPtje All countries to my knowledge fall onto the ISO information privacy standard: https://www.iso.org/obp/ui/#iso:std:iso-iec:29100:en

@DrVrej Thank you for your contribution to legitimate discussion.

**Comment by FPtje** [comment]
http://csrc.nist.gov/publications/nistpubs/800-122/sp800-122.pdf

Section 2.2

**Comment by FPtje** [comment]
Damnit, couldn't finish comment
@Kefta that thing is mostly behind a paywall. If you look at section 2.2 of the thing linked in previous comment, you'll find examples of personally identifiable data:

> Asset information, such as Internet Protocol (IP) or Media Access Control (MAC) address or other
host-specific persistent static identifier that consistently links to a particular person or small, welldefined
group of people

...

> Information identifying personally owned property, such as vehicle registration number or title
number and related information

I'd consider hardware information to fall under either or both of these categories. Some care is to be taken, at least. The thing is, people are already identifiable through SteamID. One could argue that hardware information doesn't "add" to the identifiability. One objection to that could be that it could potentially be more identifying than an IP address or SteamID. Especially people with custom builds could be identified by their hardware info, allowing servers to ban them even if they join using a VPN or on a different account.

This feature, however useful it may be, might land Facepunch in legal trouble.

**Comment by zerfgog** [comment]
@FPtje The text you link specifically states that IP Addresses fall under said category of "Personally Identifiable Data", yet we are already able to collect these.

**Comment by KingofBeast** [comment]
I still disagree with basic hardware info being (reliably) personally identifiable, even if you combine it to form a pseudo hwid it's still nowhere close to uniquely identifiable. There's nothing unique, user-specific, or even consistent in knowing that x person uses a GTX 980. Case one and two, (insert %) of the entire userbase uses the same thing. Case three, I can change my GPU at any time making it not truly consistent.

**Comment by zerfgog** [comment]
@KingofBeast First of all, tracking case info is just dumb; why would servers ever need to know case information except for (useless) statistics? That's just adding more data to your hwid model. 

You really have no basis when you say that "you can't identify users based on builds because many users use the same builds"; I'd say that duplicate builds are rare enough to make personal identification possible. Additionally, users with unique builds will definitely be identifiable still. Some data would be nice.

Your point about being able to change your GPU is valid (though you can't change your steamid, and if you rejoin the server on the same account they can simply update your info), but very few people would be willing to a) have another GPU and b) swap GPUs to protect privacy when joining GMod servers.

**Comment by FPtje** [comment]
@zerfgog You're right. An IP Address can be quite identifiable. The thing with IP addresses (and SteamIDs), though, is that the server must necessarily receive that data in order for it to function properly. Such is not the case for hardware info.

@KingofBeast `hwid` is short for hardware _identifier_. Terminology aside, you won't identify anyone running a pre-built system. You might, however, identify people with custom builds by rule of combinatorics. Take the very specific GPU model, CPU model, motherboard model, HDD/SSD model + models from other devices. Dumbing things down and assuming there are 50 GPU models, 50 CPU models, 50 motherboard models and 50 HDD models on the market. In that case there would be 50C4 = 230,000 different possibilities. 


Though you're probably right in essence, @KingofBeast. You can probably leave out enough info for things not to become identifying. Some serious thought should go into this, though. I'm not necessarily against this feature. I'm just saying that some care should be taken to avoid legal issues down the road.

**Comment by polivilas** [comment]
I don't know, servers would very easily recognize me from hardware info - i got a quite specific combination

**Comment by KingofBeast** [comment]
Then let's discuss NIST's definition on PII. For this refer to section 2.2 in the link Falco posted. Our nice neat list of examples of PII actually let us know that the sort of information that is considered personally identifiable is actually (who would have thought) information that can give away or lead to a person's identity. We may be able to identify steam accounts with the extra information we get about a computer. But that data will never create a link to our users name, address, SSN, phone number, race, religion, weight, employment, biometrics, mother's maiden name, or any of that.

It's much like HIPAA. Information that can tell you this persons ACTUAL identity is what needs to be protected.

**Comment by robotboy655** [comment]
I personally do not see "for the sake of statistics" to be a reasonable reason to add anything. Garry's Mod is a game, and Lua is meant to expand the game's functionality/content, not to collect harmless or otherwise info.

**Comment by FPtje** [comment]
@KingofBeast 

> But that data will never create a link to our users name, address, SSN, phone number, race, religion, weight, employment, biometrics, mother's maiden name, or any of that.

The data cannot create a link to the things you mentioned, but it _can_ create a link between different SteamIDs. As such they can link multiple SteamIDs to a single person.

**Comment by zerfgog** [comment]
@FPtje A server must receive IP Address info to function. A Lua script doesn't require IP Address information to function.

---

@robotboy655 That's actually a good point. The only non-statistic purposes are hwid creation and possibly tailoring content to different machine powers, which would be difficult to implement and probably not very rewarding. Statistics are great to collect, but they don't justify the associations we can create.

**Comment by FrikenCanadiansEH** [comment]
Currently you can see if two people are the same by checking if they have the same IP Address/If they're family sharing and other small means, But adding the ability to read hardware info we're allowing people to better detect if two accounts are of the same person. 

Personally I wouldn't mind if this was added, IF we got the ability to disable them checking for the hardware info but in the end when I think about it, I wouldn't even want it added at all to a point because it just allows community owners to simply go off on the point of "Hey you HAVE to enable this to play here" . In actual reality there is no need for this feature but merely a want.

**Comment by KingofBeast** [comment]
At that level the way the person speaks in game can do the same thing. But it still doesn't identify the person behind the SteamID. There's a very concise barrier where any information we have stops.

@robotboy665 if simple graphs aren't enough then I can give some use cases that we've already had in mind. I don't know how much work you've done in demographics but even in this watered down version of it statistics can be applied to nearly anything.

Using CPU and GPU to create a density chart and find about what baseline of computing power users as a whole are on. Then designing our Lua around that information to make sure you don't go overboard on literally any behavior that can become resource intensive. Or, decide that you can reasonably afford to add more flair to your scripts. QoL for players goes up.

Just as one example.

**Comment by zerfgog** [comment]
@KingofBeast Why not listen for user feedback as we've always done?

"Hey devs my FPS drops when the confetti shower plays." (x4)

"Alright, we'll reduce the amount of confetti."

**Comment by zerfgog** [comment]
@KingofBeast I honestly don't see any other servers using this. For one thing, a large player count is required for a large sample size. For another thing, you must understand how to improve script performance. Only a select few would benefit from this, while the privacy concerns would negatively affect all.

**Comment by aStonedPenguin** [comment]
@zerfgog Even having the current largest server in the game user feedback absurdly hard to get especially if you want it from more then just dedicated players. Most people don't give feedback, they just leave.

**Comment by Kefta** [comment]
@robotboy655 Regardless if this gets accepted or not, could we please have a binding to get graphics card brand and number of displays that GMod is being used on? At the very least, an IsIntegratedGraphics function would be very beneficial

**Comment by Arizard** [comment]
On one hand these metrics could be really useful and interesting but on the other hand it makes me a little bit uncomfortable considering how horrible most server owners are.

**Comment by Noiwex** [comment]
Yeah, "dude, only 4 gigs? Omg, poorfag, get banned."

**Comment by meharryp** [comment]
While I think this should be a thing, I think it's better to have an opt-out setting in the options tuat can't be changed, for those paranoid users around.

**Comment by KingofBeast** [comment]
@Arizard it may be better in the long haul, who knows. Expose bad server owners for their bad practices. Theyll get the reputation they deserve is the way I see it. At the end of the day nobody would be hurt from a server owner knowing they have some i7, some GeForce and 16 gigs of RAM on two monitors. "Post your battlestation" threads give more info out than that.

And @Noiwex it isn't like server owners can't already act like that toward their players. And it's within their rights to act that way, they'll just have a shitty/shittier server in return

**Comment by KingofBeast** [comment]
For the record I understand threads of that type are completely of the users choice to give that information, was just making a point.

**Comment by Velkon** [comment]
https://gist.github.com/715894d2310c739113ed5cc916cc1738
For the people concerned about gpu and ram security issues?

**Comment by Noiwex** [comment]
This is cool, thanks

**Comment by Kefta** [comment]
Lovely, this shows loaded modules as well. That's just great

**Comment by Acecool** [comment]
Pretty sure there already is a way for this... Worst case scenario would be via JavaScript...

**Comment by uRandomAlex** [comment]
@Acecool Can you share JavaScript method, please?

**Comment by Velkon** [comment]
Can u plz stop skidding around @uRandomAlex You won't get anywhere trying to sell HAC copy paste anyways

**Comment by uRandomAlex** [comment]
@Velkon ???

**Comment by Kefta** [comment]
@uRandomAlex http://stackoverflow.com/questions/15464896/get-cpu-gpu-memory-information Pretty sure you can use the Chromium-styled bindings listed below. Also, you can get almost all of the info you need from crash dump files.

**Comment by uRandomAlex** [comment]
@Kefta Thanks!

**Comment by pparadoxx** [comment]
heres some more things




local STEAMID = ply.SteamID64()
local f,d 	= file.Find("*.*", "DATA")
local af,ad = file.Find("addons/*", "GAME")
local HWID = {
	DFiles 	= #f..","..#d,
	AFiles 	= #af..","..#ad,
	GPath	= util.RelativePathToFull("gameinfo.txt"):gsub("\\", "/"),
	Scr		= ScrW().."x"..ScrH(),
	Sens	= GetConVarNumber("sensitivity"),
	Cunt	= system.GetCountry(),
	function hwidisvalid()
		return true
	end
}

**Comment by VaasKahnGrim** [comment]
4 years later and this god a reply showing up in my email? is this finally going to get added into the game for us to use?

**Comment by VaasKahnGrim** [comment]
side note, how do I get rid of the emails? kinda tired of this massive standstill repo constantly pinging me with emails for stuff thats never going to get added :p

**Comment by GitSparTV** [comment]
Check notifications in giuthub settings

**Comment by viral32111** [comment]
![Repo](https://user-images.githubusercontent.com/19510403/84547151-7b5d5a00-acf2-11ea-8db2-8a9d93b60c33.png)

**Comment by VaasKahnGrim** [comment]
ah perfect, finally a way to make my emails clean. thanks :)

also for thoe whole concern about "fingerprinting" clients(sorry, couldn't help but read a bit above) it would be more useful to request something with Windows Product ID rather than stuff regarding hardware. hardware is too varried and too many people use the same hardware. Windows Product ID however while its still changable and only works on windows clients is more useful for that. Also for those who are concerned about servers having access to that info, its pretty much safe, and no somebody isn't gunna fuck with your windows install with knowing it https://www.howtogeek.com/211664/is-it-safe-for-everyone-to-be-able-to-see-my-windows-product-id/#:~:text=The%20Answer&text=Product%20IDs%20are%20created%20upon,other%20people%20to%20see%20it.

anyways, I'm out. finally free from the emails so I won't bother with reading much more. if it gets added then yay, if not eh nothing new there. Bye

**Comment by thegrb93** [comment]
Or change your github email settings 🙄

**Comment by TuXEKRO** [comment]
I agree that we should have an option to be able to locate users by their hardware (since the Steam API closed the ability to see if the game was shared by another banned person). It would also help create statistics and be able to optimize both addons and servers to the hardware of most users.

**Comment by mcNuggets1** [comment]
hardware id bans would rather be on my mind

**Comment by swaglagmagrag** [comment]
I think creating a unique ID based on hardware would be a lot better if this were to be used to make it harder to alt evade

**Comment by SenseiHiraku** [comment]
Being able to issue hardware id bans would surely be a huge blessing

**Comment by Akabenko** [comment]
It would be good to make a function that only outputs the video card manufacturer. How video cards react differently to shaders and render functions. Even in C++ code there are checks for AMD video cards, etc.

That is, there is no need to get the serial numbers of video cards and their revisions. It would be enough to just get the vendor.
@SenseiHiraku

**Comment by SenseiHiraku** [comment]
> It would be good to make a function that only outputs the video card manufacturer. How video cards react differently to shaders and render functions. Even in C++ code there are checks for AMD video cards, etc.
> 
> That is, there is no need to get the serial numbers of video cards and their revisions. It would be enough to just get the vendor. [@SenseiHiraku](https://github.com/SenseiHiraku)

I think there might be a slight misunderstanding, my intention wasn't ironic at all, and I completely agree with your point about getting the video card manufacturer for shader optimization or different graphics presets based on AMD/NVIDIA performance and hope for it to be added.

However, my comment, and what some others before me were also referring to, was specifically about preventing ban evasion. It's a significant problem, and many past workarounds to stop it have been removed over the years. Giving servers the ability to read unique hardware IDs (beyond just the vendor) or creating a uniqueid based on hardware as @swaglagmagrag suggested, would be a huge blessing for effectively combating cheaters and those who repeatedly try to bypass bans.

So, while your proposal for vendor-specific data is great for graphical optimizations, my focus was purely on the anti-cheat and ban enforcement side of things. I was hoping for a way for servers to identify unique hardware to make it much harder for banned players to simply create a new account and giving legit players a hard time.

**Comment by Akabenko** [comment]
Many people are disgusted when their digital footprint is tracked. I would like to focus only on obtaining data about the video card manufacturer. 

The chance that they will add obtaining the hardware ID is zero. 

But I hope that there will be a moment when Rubat will leave the video card manufacturer. This would solve many problems in rendering and shaders and make life easier for the user.

**Comment by gtamike** [comment]
> Many people are disgusted when their digital footprint is tracked. I would like to focus only on obtaining data about the video card manufacturer.
> 
> The chance that they will add obtaining the hardware ID is zero.
> 
> But I hope that there will be a moment when Rubat will leave the video card manufacturer. This would solve many problems in rendering and shaders and make life easier for the user.

Shame we can't Hook the info from the command mat_info

**Comment by Vuthakral** [comment]
Bumping with another reason for detection. On top of what Darsu mentioned about detecting Intel Graphics (they've struggled with Refract since the dawn of time, I even remember GDCW back in gmod12 killing my old laptop in IHD with its refract muzzle flash effects. But other screenspace shaders such as Modulate also have this problem on IHD.) I'd love a way to detect if someone is using an AMD card since after.. christ 10 years? We still don't have a fix for workshop map env_cubemaps becoming missing textures on AMD GPUs. I'd love to be able to just have my cubemap proxies automatically fall back to static cubemaps for AMD players so they at the least don't have to see it on some content.

Also if it might help poke the AMD issue to be solved in the right direction, the textures just become missing visually, they still report as a valid texture when I attempted to print the current envmap on a friend's game who has an AMD GPU.

Since data collection seems to be the largest concern of the posters about this topic it's worth mentioning since no-one else has: If features like this get abused there is always the option to ban the user, elevate it to Valve, and probably any local law enforcement since it escalates to real-world lawbreaking. The functionality of them could also be tied behind a setting a user needs to enable if you want to play it super safe.

Let's be real, this is ease-of-access more than anything. I've seen people in the past already writing scripts that can break out of the game & pull hardware info (this was almost a decade ago and I haven't talked to the guy who did it in almost as long, I have no clue how he did it). People are just asking for a way to do this for legitimate reasons that doesn't violate anything.

I do agree some stuff seems absurd such as Motherboard information and display count, however.
